addons/web/static/src/core/file_viewer/file_model.js

```diff
@@ -120,7 +120,7 @@ export const FileModelMixin = T => class extends T {
      * @returns {string}
      */
     get urlRoute() {
-        if (this.isUrl) {
+        if (this.isUrl && !this.isViewable) {
             return this.url;
         }
         if (this.uploading && this.tmpUrl) {
```

Comment:
Ok for fix.
The route "web/image" and `web/content` below are implicitly linked to non-uploading files that are viewable, so the code would feel more natural with `isViewable` next to `/web/image` & `/web/content` branching.

Reply:
I am not sure I understand the comment. you mean put extra check for `isViewable` below?

Reply:
code would be easier to read as follow:

```
- if attachment uploading and has tmpUrl => use tmpUrl
- if attachment is viewable => use viewable urls
- otherwise => use attachment url
```

rather than:

```
- if attachment is url and not viewable => use attachment url
- if attachment uploading and has tmpUrl => use tmpUrl
- otherwise => use viewable urls
```

In the re-ordered logic, every condition makes sense by themselves.
With the current code, 1st condition refers to last condition, last condition is an opinionated way to get url but it feels off to be the fallback / default behaviour.
This contributes to making internal code harder to understand, and it doesn't help that the getter is named `urlSource` which is not that great to get a grasp how it's different from `url` without looking at the code. Probably should be renamed to `viewUrl`?

Anyway this are quite minor comments for a fix.

Reply:
I think the problem here is
`- otherwise => use attachment url`
should not be a default behavior as well.
So I would still prefer to use the viewer route as default.

Reply:
The code will be
```js
    /**
     * @returns {string}
     */
    get urlRoute() {
        if (this.uploading && this.tmpUrl) {
            return this.tmpUrl;
        }
        if (this.isViewable) {
            return this.isImage ? `/web/image/${this.id}` : `/web/content/${this.id}`;
        }
        if (this.isUrl) {
            return this.url;
        }
        return ??; // what is the code here?
    }
    ```

Reply:
I think you are overthinking. `viewUrl` fallback to `url` makes sense when we can use a shiny url. This is a good last resort :D

Reply:
oh, url is only for `url` typed attachment and I don't think that is the common case.

Reply:
That's because we redirect viewable content mostly with `/web/image/` & `/web/content/`, but this makes sense that we use the raw url if there's no `viewUrl`/`urlRoute`.

If the content is not viewable makes sense that we just use the url. This is not embedded-ly viewable inside odoo but there's at the very least a showing of URL to view the file in its original context, i.e. clicking on the url.

Reply:
I'm pretty sure Zelong is right here, we need to fallback on /web/image (or content), not on url.
url is more often than not empty, but the route with id is always "possible".

I'm wondering if we ever need to use url at all actually.

Reply:
The purpose of this `urlRoute`, which again is very poorly named, is to provide the URL for viewing the file inside Odoo. Whenever we consider the file viewable in odoo with `isViewable`, then it uses `/web/image` and `/web/content`.

Using the `url` field of the file should be a last resort, because this means we simply use the stored URL of the file, which is very likely not great for viewing the file inside Odoo but oh well that's why this should be the fallback when everything else fails.

Reply:
The odoo url with attachment id always redirect to the actual URL when it is necessary.
All attachments have id, so the odoo url always works.
Many attachments don't have url. It makes no sense to use that as fallback.

Also `urlRoute` is not only used for viewable content (in `defaultSource`), but also for fetching content in several flows (`onClickUseAsAttachment`/`onClickAttachToMessage` in knowledge, `getImageUrl` in attachment list, `fetchFile` in voice player

Reply:
I'm surprised there was no test to cover it, but using `url` for non-viewable non-url attachments effectively breaks the download link: it's empty so it just downloads the current page (/web in backend).

Reply:
As I suspected the attachment url does not need to be used at all, as the odoo server will properly direct based on id.
So the url part can simply be removed. Before the other PR from Zelong, the url was simply never known in JS and never used.
I'm just updating the PR to remove the condition completely.